x/oracle/keeper/historic_price.go

```diff
@@ -106,50 +80,126 @@ func (k Keeper) GetMedianDeviation(
+	prices := k.historicPrices(ctx, denom)
+	if len(prices) == 0 {
+		return false, sdkerrors.Wrap(types.ErrNoHistoricPrice, fmt.Sprintf("denom: %s", denom))
 	}
+	price := prices[0]
 
-	medianDeviation, err := k.GetMedianDeviation(ctx, denom)
+	medianDeviation, err := k.HistoricMedianDeviation(ctx, denom)
 	if err != nil {
 		return false, err
 	}
 
 	return price.Sub(median).Abs().LTE(medianDeviation), nil
 }
 
-// setMedianDeviation sets a given denom's standard deviation around
-// its median price in the last prune period since the current block.
-func (k Keeper) calcAndSetMedianDeviation(
+// calcAndSetHistoricMedianDeviation calculates and sets a given denom's standard
+// deviation around its median price in the current block.
+func (k Keeper) calcAndSetHistoricMedianDeviation(
 	ctx sdk.Context,
 	denom string,
 	median sdk.Dec,
 	prices []sdk.Dec,
-) {
-	medianDeviation := medianDeviation(median, prices)
-	k.SetMedianDeviation(ctx, denom, medianDeviation)
+) error {
+	medianDeviation, err := decmath.MedianDeviation(median, prices)
+	if err != nil {
+		return sdkerrors.Wrap(err, fmt.Sprintf("denom: %s", denom))
+	}
+
+	block := uint64(ctx.BlockHeight())
+	k.SetHistoricMedianDeviation(ctx, denom, block, medianDeviation)
+	return nil
 }
 
-func (k Keeper) SetMedianDeviation(
+func (k Keeper) SetHistoricMedianDeviation(
 	ctx sdk.Context,
 	denom string,
+	blockNum uint64,
 	medianDeviation sdk.Dec,
 ) {
 	store := ctx.KVStore(k.storeKey)
 	bz := k.cdc.MustMarshal(&sdk.DecProto{Dec: medianDeviation})
-	store.Set(types.KeyMedianDeviation(denom), bz)
+	store.Set(types.KeyMedianDeviation(denom, blockNum), bz)
+}
+
+// MedianOfHistoricMedians calculates and returns the median of the last stampNum
+// historic medians.
+func (k Keeper) MedianOfHistoricMedians(
+	ctx sdk.Context,
+	denom string,
+	numStamps uint64,
+) (sdk.Dec, error) {
+	medians := k.HistoricMedians(ctx, denom, numStamps)
+	median, err := decmath.Median(medians)
+	if err != nil {
+		return sdk.ZeroDec(), sdkerrors.Wrap(err, fmt.Sprintf("denom: %s", denom))
+	}
+	return median, nil
+}
+
+// AverageOfHistoricMedians calculates and returns the average of the last stampNum
+// historic medians.
+func (k Keeper) AverageOfHistoricMedians(
+	ctx sdk.Context,
+	denom string,
+	numStamps uint64,
+) (sdk.Dec, error) {
+	medians := k.HistoricMedians(ctx, denom, numStamps)
+	average, err := decmath.Average(medians)
+	if err != nil {
+		return sdk.ZeroDec(), sdkerrors.Wrap(err, fmt.Sprintf("denom: %s", denom))
+	}
+	return average, nil
+}
+
+// MaxOfHistoricMedian calculates and returns the maximum value of the last stampNum
+// historic medians.
+func (k Keeper) MaxOfHistoricMedians(
+	ctx sdk.Context,
+	denom string,
+	numStamps uint64,
+) (sdk.Dec, error) {
+	medians := k.HistoricMedians(ctx, denom, numStamps)
+	max, err := decmath.Max(medians)
+	if err != nil {
+		return sdk.ZeroDec(), sdkerrors.Wrap(err, fmt.Sprintf("denom: %s", denom))
+	}
+	return max, nil
+}
```

Comment:
https://github.com/umee-network/umee/pull/1631 Let's leave them in for now since that is what we already decided over here: https://github.com/umee-network/umee/pull/1631

Reply:
It should be avoided. User should not call `MedianOfMedian` or `MaxMedian` through the oracle keeper, because it will cause to fetch all medians twice. So at minimum, if we want to keep it, we should clearly describe it in the function comment.